eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/ChannelRelay.scala

```diff
@@ -190,58 +189,68 @@ class ChannelRelay private(nodeParams: NodeParams,
+            case (channel, _: RelaySuccess) if channel.commitments.availableBalanceForSend > r.payload.amountToForward => channel
           }
           .toList // needed for ordering
           // we want to use the channel with:
           //  - the lowest available capacity to ensure we keep high-capacity channels for big payments
           //  - the lowest available balance to increase our incoming liquidity
-          .sortBy { case (_, commitments) => (commitments.capacity, commitments.availableBalanceForSend) }
+          .sortBy { channel => (channel.commitments.capacity, channel.commitments.availableBalanceForSend) }
           .headOption match {
-          case Some((preferredShortChannelId, commitments)) if preferredShortChannelId != requestedShortChannelId =>
-            context.log.info("replacing requestedShortChannelId={} by preferredShortChannelId={} with availableBalanceMsat={}", requestedShortChannelId, preferredShortChannelId, commitments.availableBalanceForSend)
-            Some(preferredShortChannelId)
-          case Some(_) =>
-            context.log.debug("requested short channel id is our preferred channel")
-            Some(requestedShortChannelId)
-          case None if !alreadyTried.contains(requestedShortChannelId) =>
-            context.log.debug("no channel seems to work for this payment, we will try to use the requested short channel id")
-            Some(requestedShortChannelId)
+          case Some(channel) =>
+            if (requestedChannelId_opt.contains(channel.channelId)) {
+              context.log.debug("requested short channel id is our preferred channel")
+              Some(channel)
+            } else {
+              context.log.info("replacing requestedShortChannelId={} by preferredShortChannelId={} with availableBalanceMsat={}", requestedShortChannelId, channel.channelUpdate.shortChannelId, channel.commitments.availableBalanceForSend)
+              Some(channel)
+            }
           case None =>
-            context.log.debug("no channel seems to work for this payment and we have already tried the requested channel id: giving up")
-            None
+            val requestedChannel_opt = requestedChannelId_opt.flatMap(channels.get)
+            requestedChannel_opt match {
+              case Some(requestedChannel) if alreadyTried.contains(requestedChannel.channelId) =>
+                context.log.debug("no channel seems to work for this payment and we have already tried the requested channel id: giving up")
+                None
+              case _ =>
+                context.log.debug("no channel seems to work for this payment, we will try to use the one requested")
+                requestedChannel_opt
+            }
         }
-      case _ => Some(requestedShortChannelId) // we don't have a channel_update for this short_channel_id
+      case _ => requestedChannelId_opt.flatMap(channels.get) // we don't have a channel_update for this short_channel_id
```

Comment:
It's a bit confusing to make it look like we could return a result here whereas it will always be `None` (because we reach that case if `channels` is empty, maybe it's worth being more explicit?

```suggestion
      case _ => None // we don't have a channel_update for this short_channel_id
```

Reply:
Actually we can remove the whole pattern matching I believe. Done in 848062ff9.